# Módulo 5 - Funções

## Aulas 5.2 - Funções Customizadas e 5.4 Funções ```lambda```

### Q1.

- Sabendo que o código a seguir calcula o fatorial de n, escreva uma função chamada ```fatorial()``` que recebe um inteiro ```n``` como parâmetro e retorna o resultado do fatorial de ```n```. 
- No programa principal, peça ao usuário o valor de ```n```, chame a sua função e imprima o retorno.

```python
fat = 1
for i in range(1, n+1):
    fat *= i 
```

In [147]:
def fatorial(n):
    if n < 0:
        return "Erro: número negativo"
    elif n == 0:
        return 1
    else:
        fat = 1
        for i in range(1, n + 1):
         fat = fat * i
        return fat
    

In [148]:
num = int(input("Digite um número: "))
resultado = fatorial(num)
print(f"O fatorial de {num} é {resultado}")

O fatorial de 123 é 12146304367025329675766243241881295855454217088483382315328918161829235892362167668831156960612640202170735835221294047782591091570411651472186029519906261646730733907419814952960000000000000000000000000000


### Q2.

- Escreva uma função em Python chamada ```soma_quadrados``` que recebe dois números como parâmetros e retorna a soma dos seus quadrados. 
- No programa principal solicite ao usuário que insira dois números e utilize a função para exibir a soma dos quadrados.

In [149]:
def soma_quadrados(numero1, numero2):
    quadrado1 = numero1 * numero1
    quadrado2 = numero2 * numero2
    soma = quadrado1 + quadrado2
    return soma 


In [150]:
numero1 = int(input("Digite o primeiro número "))
numero2 = int(input("Digite o segundo número "))
resultado = soma_quadrados(numero1, numero2)
print("A soma dos quadrados é:", resultado)

A soma dos quadrados é: 32818


### Q3.

- Crie uma função em Python chamada ```soma_digitos``` que recebe um número inteiro como parâmetro e retorna a soma dos seus dígitos. Por exemplo, para o número 123, a função deve retornar 6, $(1 + 2 + 3)$.
    - O desafio aqui é separar os dígitos de um número inteiro usando operações aritméticas <br> <br>

- No programa principal solicite ao usuário que insira um número e utilize a função ```soma_digitos``` para calcular e exibir a soma dos seus dígitos.

In [163]:
def soma_digitos(numero):
    soma = 0
    for digito in str(numero):
        soma += int(digito)
    return soma


In [164]:
numero = int(input("Digite um número: "))
resultado = soma_digitos(numero)
print("A soma dos digitos é:", resultado)


A soma dos digitos é: 6


### Q4.

- Crie a função ```inverteValor()``` que recebe um inteiro de qualquer tamanho e retorna esse valor invertido usando apenas operações aritméticas
- Crie a função ```verificaInverso()``` que recebe o valor original e o valor invertido e retorna verdadeiro se ambos forem igualmente par ou igualmente ímpar. Retorne falso caso contrário.
- No programa principal, peça um valor do usuário e imprima o retorno de ambas as funções.


In [167]:
def inverteValor(numero):
    invertido = 0
    while numero != 0:
        digito = numero % 10
        invertido = invertido * 10 + digito
        numero = numero // 10
    return invertido

def verificaInverso(original, invertido):
    if (original % 2 == invertido % 2):
        return True
    else:
        return False




In [168]:
numero = int(input("Digite um número: "))
invertido = inverteValor(numero)
verificacao = verificaInverso(numero, invertido)

print("Número invertido:", invertido)
print("Verificação se ambos são pares ou ímpares:", verificacao)


Número invertido: 4321
Verificação se ambos são pares ou ímpares: False


### Q5.

Você está desenvolvendo um programa para auxiliar em cálculos de geometria básica. Crie as seguintes funções:
- A função ```calcula_perimetro_triangulo()``` que recebe três inteiros correspondentes aos lados de um triângulo e retorna o perímetro do triângulo, ou seja, a soma dos seus lados.
- A função ```calcula_perimetro_circulo()``` que recebe um inteiro referente ao raio do círculo e retorna o perímetro do círculo, dado por $2 \pi r$. Use a constante $\pi$ da biblioteca ```math```.
- A função ```calcula_perimetro_retangulo()``` que possui um parâmetro obrigatório ```lado1``` e um opcional ```lado2```, ambos inteiros. Se o valor opcional não for fornecido, significa que se trata de um quadrado. Sua função deve calcular e retornar o perímetro do retângulo, ou seja, a soma de seus lados. 
    - Para o quadrado, é dado por $4 \times lado1$
    - Para o retângulo é dado por $2 \times lado1 + 2 \times lado2$
    
- No programa principal apresente um menu com as opções disponíveis do seu sistema e uma quarta opção ```Sair```. Solicite ao usuário a opção desejada, solicite as entradas correspondentes à opção escolhida, invoque a respective função e apresente o seu retorno. Seu programa deve retornar ao menu até que o usuário escolha a opção ```Sair```

Exemplo de interação:
```
1 - Calcular perímetro triângulo
2 - Calcular perímetro círculo
3 - Calcular perímetro retângulo
4 - Sair

Opção: 1
Digite os três lados do triângulo:
3
4
5
O perímetro é: 12

1 - Calcular perímetro triângulo
2 - Calcular perímetro círculo
3 - Calcular perímetro retângulo
4 - Sair

Opção: 3
Informe os dois lados do retângulo. Se for um quadrado, digite 0 para o segundo valor:
5
0
O perímetro é: 20

1 - Calcular perímetro triângulo
2 - Calcular perímetro círculo
3 - Calcular perímetro retângulo
4 - Sair

Opção: 4
```

In [2]:
import math

def perimetro_triangulo(lado1, lado2, lado3,):
    return lado1 + lado2 + lado3
def perimetro_circulo(raio):
    return 2 * math.pi * raio
def perimetro_retangulo(lado1, lado2=None):
    if lado2:
        return 2 * (lado1 + lado2)
    return 4 * lado1

In [7]:
while True:
    print("\nEscolha uma opção:")
    print("1 - Perímetro triângulo")
    print("2 - Perímetro círculo")
    print("3 - Perímetro retângulo ou quadrado")
    print("4 - Sair")

    opcao = input("Opção: ")

    if opcao == '1':
        lado1 = float(input("Digite lado 1: "))
        lado2 = float(input("Digite lado 2: "))
        lado3 = float(input("Digite lado 3: "))
        print("Perímetro do triângulo:", perimetro_triangulo(lado1, lado2, lado3))

    elif opcao == '2':
        raio = float(input("Digite o raio do círculo:" ))
        print("Perímetro do círculo:", perimetro_circulo(raio))

    elif opcao == '3':
        lado1 = float(input("Digite o lado 2 ou pressione enter se for quadrado: "))
        if lado2:
            lado2 = float(lado2)
            print("Perímetro do retângulo:", perimetro_retangulo(lado1, lado2))

        else:
            print("Perímetro do quadrado:", perimetro_retangulo(lado1))

    elif opcao == '4':
        print("Saindo programa...")
        break
    else:
        print("Opção inválida. Tente novamente.")


Escolha uma opção:
1 - Perímetro triângulo
2 - Perímetro círculo
3 - Perímetro retângulo ou quadrado
4 - Sair
Perímetro do triângulo: 12.0

Escolha uma opção:
1 - Perímetro triângulo
2 - Perímetro círculo
3 - Perímetro retângulo ou quadrado
4 - Sair
Perímetro do círculo: 18.84955592153876

Escolha uma opção:
1 - Perímetro triângulo
2 - Perímetro círculo
3 - Perímetro retângulo ou quadrado
4 - Sair
Perímetro do retângulo: 20.0

Escolha uma opção:
1 - Perímetro triângulo
2 - Perímetro círculo
3 - Perímetro retângulo ou quadrado
4 - Sair
Saindo programa...


## Funções ```lambda```

### Q6.

Crie uma função lambda para verificar se um número é par ou ímpar. Em seguida, solicite ao usuário um número indefinidos de valores (até que o usuário digite 0). Para cada valor de entrada, informe se é par ou ímpar.

Exemplo de interação:
```
Digite os valores que deseja verificar a paridade (digite 0 para finalizar a entrada de dados):
3
ímpar
8
par
12
par
5
ímpar
7ímpar
```

In [13]:
par_ou_impar = lambda x: "par" if x % 2 == 0 else "impar"

print("Digite os valores que deseja verificar se é par ou ímpar (0 para finalizar): ")

while True:
    valor = int(input())
    if valor == 0:
        break
    print(valor)
    print(par_ou_impar (valor))

Digite os valores que deseja verificar se é par ou ímpar (0 para finalizar): 
3
impar
8
par
12
par
5
impar


### Q7.

Escreva um programa que pergunte ao usuário qual operação ele deseja: maior ou menor.
Em seguida leia uma quantidade indefinida de valores do usuário, até que o usuário digite o valor zero.
Apresente ao final o maior ou menor dos valores digitados de acordo com a escolha do usuário.

**Sua solução deve incluir pelo menos uma função ```lambda```** 

Exemplo de interação:
```
Opções: (1) maior ou (2) menor?
Opção: 1

Digite os valores de entrada. Digite 0 para finalizar a entrada de valores.
4
12
15
1
0

O maior valor é: 15
```

In [11]:
opcao = int(input("Opções: (1) maior ou (2) menor?\nOpção: "))

operacao = (lambda lista: max(lista)) if opcao == 1 else (lambda lista: min(lista))

valores = []

print("\nDigite os valores de entrada. Digite 0 para finalizar a entrada de valores: ")

while True:
    valor = int(input())
    if valor == 0:
        break
    valores.append(valor)

if valores:
    resultado = operacao(valores)
    if opcao == 1:
        print(f"\nO maior valor é: {resultado}")
    else:
        print(f"\nO menor valor é: {resultado}")
else:
    print("Nenhum valor foi inserido>")


Digite os valores de entrada. Digite 0 para finalizar a entrada de valores: 

O maior valor é: 15
